allennlp/predictors/coref.py

```diff
@@ -70,6 +72,55 @@ def predict_tokenized(self, tokenized_document: List[str]) -> JsonDict:
         """
         instance = self._words_list_to_instance(tokenized_document)
         return self.predict_instance(instance)
+    
+    @staticmethod
+    def replace_corefs(document: Doc, clusters: List[List[int]]) -> str:
+        """
+        Uses a list of coreference clusters to convert a spacy document into a 
+        string, where each coreference is replaced by its main mention. 
+        """
+    
+        resolved = list(tok.text_with_ws for tok in document)
+        
+        for cluster in clusters:
+            mention_start, mention_end = cluster[0][0], cluster[0][1] + 1
+            mention_span = document[mention_start: mention_end]
+            
+            for coref in cluster[1:]:
+                final_token = document[coref[1]]
+                # Correctly formats possessive mentions with 's endings
+                # These include my, his, her, as well as persons's, computer's, etc.
+                if final_token.tag_ in ["PRP$", "POS"]:
+                    resolved[coref[0]] = mention_span.text + "'s" + final_token.whitespace_
+                else:
+                    resolved[coref[0]] = mention_span.text + final_token.whitespace_
+                for i in range(coref[0] + 1, coref[1] + 1):
+                    resolved[i] = ""
```

Comment:
What is this doing exactly?

Reply:
Since `resolved` is a list of the tokens in the document with white space, line 96 replaces the first coreference token with the mention it refers to, while lines 97-98 goes through and masks out all subsequent tokens with `""`, so that they are ultimately eliminated in the `"".join` that follows. This procedure is necessary for replacing multi-word coreferences with a single mention, e.g. "the country" with "China" in example. This logic is borrowed from lines 262-271 [here](https://github.com/huggingface/neuralcoref/blob/master/neuralcoref/neuralcoref.pyx).

Reply:
Ah--that makes sense.  I recommend turning part of your explanation into a comment.

Reply:
Sounds good, yea I'll add that comment. `resolved.remove(i)` work work as well, but it's slightly more inefficient (quadratic vs. linear).